File: dodam-system-domain/dodam-domain-rds/src/main/java/b1nd/dodam/domain/rds/member/repository/MemberRepository.java
package b1nd.dodam.domain.rds.member.repository;

import b1nd.dodam.domain.rds.member.entity.Member;
import b1nd.dodam.domain.rds.member.enumeration.ActiveStatus;
import b1nd.dodam.domain.rds.member.enumeration.MemberRole;
import b1nd.dodam.domain.rds.member.exception.MemberNotFoundException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

public interface MemberRepository extends JpaRepository<Member, String> {

    Optional<Member> findByIdAndPw(String id, String pw);

    List<Member> findByCreatedAtAfter(LocalDateTime createdAt);

    @Query("SELECT m FROM member m LEFT JOIN student s ON m.id = s.member.id " +
            "WHERE m.status = :status " +
            "ORDER BY CASE WHEN s.id IS NOT NULL THEN 0 ELSE 1 END ASC, " +
            "s.grade ASC, s.room ASC, s.number ASC")
    List<Member> findByStatusOrderByStudent(ActiveStatus status);

    @Query("""
        SELECT m FROM member m
        LEFT JOIN student s ON s.member = m
        WHERE (:keyword IS NULL OR m.name LIKE %:keyword%) 
        AND (:grade IS NULL OR s.grade = :grade)
        AND (:role IS NULL OR m.role = :role)
        AND (:status IS NULL OR m.status = :status)
    """)
    Page<Member> searchMembers(
            @Param("keyword") String keyword,
            @Param("grade") Integer grade,
            @Param("role") MemberRole role,
            @Param("status") ActiveStatus status,
            Pageable pageable
    );

    List<Member> findByNameContains(String name);

    default Member getById(String id) {
        return findById(id)
                .orElseThrow(MemberNotFoundException::new);
    }


    List<Member> findByModifiedAtAfter(LocalDateTime modifiedAt);

}

File: services/service-member/src/main/kotlin/com/b1nd/dodamdodam/member/domain/member/repository/MemberRepository.kt
package com.b1nd.dodamdodam.member.domain.member.repository

import com.b1nd.dodamdodam.core.security.passport.enumerations.RoleType
import com.b1nd.dodamdodam.member.domain.member.entity.MemberEntity
import com.b1nd.dodamdodam.member.domain.member.enumeration.ActiveStatus
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface MemberRepository : JpaRepository<MemberEntity, Long> {
    fun findByUsername(username: String): MemberEntity?
    fun existsByUsername(username: String): Boolean
    fun existsByEmail(email: String): Boolean
    fun findByStatus(status: ActiveStatus): List<MemberEntity>

    @Query("SELECT m FROM MemberEntity m JOIN MemberRoleEntity mr ON mr.member = m WHERE mr.role = :role AND m.status = :status AND m.name LIKE %:name%")
    fun searchByNameAndRoleAndStatus(
        @Param("name") name: String,
        @Param("role") role: RoleType,
        @Param("status") status: ActiveStatus,
        pageable: Pageable
    ): Page<MemberEntity>

    @Query("SELECT DISTINCT m FROM MemberEntity m LEFT JOIN MemberRoleEntity mr ON mr.member = m WHERE (:name IS NULL OR m.name LIKE %:name%) AND (:role IS NULL OR mr.role = :role) AND (:status IS NULL OR m.status = :status)")
    fun searchMembers(
        @Param("name") name: String?,
        @Param("role") role: RoleType?,
        @Param("status") status: ActiveStatus?,
        pageable: Pageable
    ): Page<MemberEntity>
}
